### ISSUE TITLE
redirect: Clean file URLs as well

### ISSUE THREAD

**Issue by LecrisUT** [readthedocs/readthedocs.org#11220]
## What's the problem this feature will solve?

Currently the clean URL redirects `/path/to/my_page/` to `/path/to/my_page.html`, but it doesn't do that for `/path/to/my_page`.

## Describe the solution you'd like

It would be good to have an equivalent that can to that redirect for the file paths as well (those that do not have any suffix).

## Alternative solutions

Tried to write a redirect manually for `/*` to `/:splat.html`, but that did not work, not to mention that it would fail for properly suffixed files.

Probably this is an issue that can/should be solved upstream, but I am not sure what solution to suggest to them.

**Comment by humitos** [comment]
> Currently the clean URL redirects `/path/to/my_page/` to `/path/to/my_page.html`, but it doesn't do that for `/path/to/my_page`.

If I understand correctly, the redirect doesn't work when there is no a trailing slash, right?

**Comment by LecrisUT** [comment]
> If I understand correctly, the redirect doesn't work when there is no a trailing slash, right?

Yes, that sums it up

> What do to mean by upstream here? Whom are them?
> I'm not sure to follow this. How is npx serve related to this issue here?

`mystmd` is the upstream site-generator. Basically is it an issue for the site-generator and/or site hosting? It seems that `npx serve` does this kind of redirection automatically, that's why I mention it here. But it could be an issue on how the pages are produced, since I don't think `sphinx` has a similar issue if you open the `index.html` file manually.

**Comment by choldgraf** [comment]
Here's my understanding of this issue, I think a more descriptive title would be "Clean URLs should redirect URLs without a trailing slash". Basically:

In the JS ecosystem it's common for servers to automatically redirect clean URLs to add `.html` if such a file exists (e.g. `foo.com/bar` to `foo.com/bar.html`).

In the "Clean URLs" feature on RTD, this seems to only exist for _directory_-style URLs (e.g. `foo.com/bar/` redirects to `foo.com/bar.html`, but `foo.com/bar` does not).

@humitos if it's helpful to chat this through at all I'm happy to do so! 

I think this is one of the main blockers for [Jupyter Book 2](https://next.jupyterbook.org) support in readthedocs. [MyST](https://mystmd.org) is the underlying documentation engine for JB2.

**Comment by choldgraf** [comment]
Here's an example of this not working. On the following site I have Clean URLs configured to go from `dirhtml/` to `page.html` and you can see how a trailing slash is _required_:

https://test-myst-rtd.readthedocs.io/en/latest/

<img width="1536" height="850" alt="Image" src="https://github.com/user-attachments/assets/4b9a8a99-e308-41c8-a5f2-361e0e10363e" />

**Comment by humitos** [comment]
> In the JS ecosystem it's common for servers to automatically redirect clean URLs to add `.html` if such a file exists (e.g. `foo.com/bar` to `foo.com/bar.html`).

Do you know a hosting platform that supports this type of redirect? If so, do you have an example of a documentation project working in this way in that hosting?

----

If this is the only change required here, we should create a new redirect type called `Clean URL, without trailing slash, to HTML (file to file.html)` so it's explicit and we don't have to change behavior on the current redirect.

**Comment by choldgraf** [comment]
The best example I can think of is netlify, which is what the Turing Way uses

https://book.the-turing-way.org/

This uses [post processing in Netlify](https://docs.netlify.com/site-deploys/post-processing/), which is enabled by default. They call it "pretty URLs".

GitHub pages also works in the same way. I think many jupyter books use GitHub pages as well, but I can't find "clean URLs" or "pretty URLs" documented in GitHub explicitly.

Interestingly it seems like netlify uses a mechanism other than redirects for this (or I assume so given that the section is about post processing and they have a different page documenting redirects)

BTW if you implement this and want us to test it out let me know and I can try with next.jupyterbook.org . Is it possible to switch this on with the yml configuration?

**Comment by humitos** [comment]
> This uses [post processing in Netlify](https://docs.netlify.com/site-deploys/post-processing/), which is enabled by default. They call it "pretty URLs".

OK, I see. They implement this as a rewrite, not as a redirect. I don't think we can go in that direction on Read the Docs because it will require more structural changes. I went ahead and [implemented a new redirect type for this use case](https://github.com/readthedocs/readthedocs.org/pull/12323) and tested it locally. It worked great!

If we move forward with this implementation, you will need to create a redirect under your project's settings, similar to how the other redirects are created.

**Comment by choldgraf** [comment]
If you give me lightweight instructions I'm happy to try them out!

It sounds like the other solution would be to change this on the myst side right? Either to have a setting for Myst to generate .html links, or to generate folders w index.html inside?

**Comment by humitos** [comment]
Using https://github.com/choldgraf/test-myst-rtd as example, I was able to make it work locally with #12323 and two redirects: `Clean URL to HTML (file/ to file.html)` and (the new) `Clean URL, without trailing slash, to HTML (file to file.html)`. This makes `/page` and `/page/` to redirect to `page.html` --which is served correctly.

**Comment by ericholscher** [comment]
When I go to https://book.the-turing-way.org/foo -- it doesn't seem to redirect me to https://book.the-turing-way.org/foo.html. Does this only happen _when_ there is an existing `foo.html` file? 

```
-> curl -IL https://book.the-turing-way.org/foo
HTTP/2 404
```

**Comment by humitos** [comment]
> When I go to [book.the-turing-way.org/foo](https://book.the-turing-way.org/foo) -- it doesn't seem to redirect me to [book.the-turing-way.org/foo.html](https://book.the-turing-way.org/foo.html). Does this only happen _when_ there is an existing `foo.html` file?

@ericholscher That's because Netlify doesn't implement this as a redirect, but as a post-process rewrite feature : https://docs.netlify.com/build/post-processing/overview/

It serves https://book.the-turing-way.org/collaboration/collaboration.html at https://book.the-turing-way.org/collaboration/collaboration as well. 


It redirects https://book.the-turing-way.org/collaboration/collaboration/ (trailing slash) to https://book.the-turing-way.org/collaboration/collaboration, tho.

```
$ curl -IL https://book.the-turing-way.org/collaboration/collaboration/
HTTP/2 301 
location: /collaboration/collaboration
server: Netlify
```

**Comment by choldgraf** [comment]
Do y'all have anybody running Docusaurus via RTD? If so, that might be something to learn from (either workarounds they found, or ways that docusaurus solves this problem). It seems that [docusaurus also does cleanurls by default](https://docusaurus.io/docs/migration/v2/manual#removed-fields).

EDIT: [This helpful guide to routing in Docusaurus](https://docusaurus.io/docs/advanced/routing#routes-become-html-files) has the answer. It sounds like they emit `pagename/index.html`...I wonder if that means that Docusaurus *wouldn't* work in RTD if the single-page routing setting was turned on.

In addition - [this trailing slash guide to hosting providers](https://github.com/slorber/trailing-slash-guide) is really nice as well. @humitos y'all might wanna try to get in there 😅

**Comment by ericholscher** [comment]
Yes, we have it working and documented on RTD with the default routing config. It works fine. The main change required is setting the base URL: https://docs.readthedocs.com/platform/stable/intro/docusaurus.html

It seems like we have a proposed path forward with the redirects PR that @humitos put together. I think that is probably the shortest way to make it work on our side.

**Comment by humitos** [comment]
Docusaurus works out of the box because they create the `/pagename/index.html` as you mentioned, @choldgraf, but also because Read the Docs automatically redirect `/pagename -> /pagename/` (adding a trailing slash), which servers `/pagename/index.html`. This doesn't work on mystmd because that SSG doesn't generate the `pagename/index.html` but `pagename.html` instead.

> If trailingSlash is set to false, the build would emit intro.html instead of intro/index.html.

_(from https://docusaurus.io/docs/advanced/routing#routes-become-html-files)_

I think if you use `trailingSlash: false` in Docusaurus, you will get the same behavior we are currently experiencing with mystmd, and we will need something like the PR I opened. I will try this and reply back here.

> In addition - [this trailing slash guide to hosting providers](https://github.com/slorber/trailing-slash-guide?rgh-link-date=2025-07-16T16%3A53%3A22.000Z) is really nice as well. [@humitos](https://github.com/humitos) y'all might wanna try to get in there 😅

Interesting! Yeah, probably a good idea to open a PR there once we close this issue 😄 

> Do y'all have anybody running Docusaurus via RTD?

I have the basic default project deployed at https://test-builds.readthedocs.io/en/docusaurus/ and it works great. That project is the one we have been using for testing different features, including addons.

**Comment by humitos** [comment]
I tested my PR locally with a mystmd example and it worked as expected. The link from the navbar "How to MyST, without being mystified 🧙" pointing to `/en/mystmd/paper` is redirected to `/en/mystmd/paper.html` and it's correctly served by Read the Docs.

The test project is deployed at https://test-builds.readthedocs.io/en/mystmd/, but it's not currently working since the PR is not deployed yet. Once deployed, we can come back to that project and confirm it works. This is the [`readthedocs.yaml`](https://github.com/readthedocs/test-builds/blob/mystmd/.readthedocs.yaml) file that builds that site. We should update our "Popular documentation tools" page once we merged this and confirm it's working as expected.

Note that I found there is still and issue with mystmd that it's outside our scope and it's reported in the addons repository:
* https://github.com/readthedocs/addons/issues/278

We can continue that conversation there, since out of the scope of this issue, but I wanted to note it here as well, since there are more work to do from the mystmd side to be fully supported on Read the Docs.

**Comment by humitos** [comment]
> I think if you use `trailingSlash: false` in Docusaurus, you will get the same behavior we are currently experiencing with mystmd, and we will need something like the PR I opened. I will try this and reply back here.

Confirmed. Docusaurus with `trailingSlash: false` _needs_ the redirect from my PR to work properly. Otherwise, it shows a not found page:

<img width="574" height="363" alt="Image" src="https://github.com/user-attachments/assets/786cc5bd-adda-4719-8c70-627a91ec7d2a" />

**Comment by choldgraf** [comment]
I think that this now works now that we've shipped this one!

- https://github.com/jupyter-book/mystmd/pull/2178

I believe that, strictly speaking, this issue is still valid, but it I don't believe it's a blocker for myst :-)

**Comment by humitos** [comment]
We have a discussion about this issue and the redirect implementation proposed in #12323. 

TLDR, we are not implementing this feature for now 😄 

These are the notes from that meeting:

- Adding another type of redirect makes the platform more complex to everybody.
- The redirect doesn't _solve_ the issue but hides the problem: the URL won't  be "clean" since it will redirect `/foo -> /foo.html`. The URL will work, but people wanting the clean URLs without the `.html` part on it, won't get that.
- Eventually, the implementation could be done at El Proxito (server) level. This will require an extra S3 call to check if `foo.html` exists before performing the system level redirect `/foo -> /foo/` that it's already in place. This will add an extra call to a bunch of requests adding extra time and costs.
- Previous extra time and costs could be reduced by making this feature under a setting instead.
- We decided to document all known tools that have a setting to output HTML files in a standard way (#12345, #12344)
